Develop

Mary-cross1296 opened the pull request.

Mary-cross1296 committed:
Тест на запуск сервера - не работает. Тест на проверку БД - работает

Mary-cross1296 committed:
Вынесен запуск сервера в отдельную функцию. Добавлена обработка сигналов.

Mary-cross1296 committed:
Правки запуска сервера

Mary-cross1296 committed:
Правки функции HttpServer, добавлена обработка ошибок и сигналов

Mary-cross1296 committed:
Черновик недоделанного задания по БД

Mary-cross1296 committed:
Функция создания базы данных - рабочий вариант

Mary-cross1296 committed:
Правка функции ChekingDataBase()

Mary-cross1296 committed:
Правки функции ChekingDataBase()

Mary-cross1296 committed:
Окончательная рабочая версия функции ChekingDataBase()

Mary-cross1296 committed:
Добавлены функция СalculatDailyTask, СalculatYearlyTask, СalculatWeeklyTask

Mary-cross1296 committed:
Работа над функциями СalculatMonthlyTask, СalculatDayOfMonthTask, СalculatDayOfMonthsTask

Mary-cross1296 committed:
Работа над функциями СalculatMonthlyTask, СalculatDayOfMonthTask, СalculatDayOfMonthsTask

Mary-cross1296 committed:
Все тесты для функции CalculatDayOfMonthTask завершаются корректно

Mary-cross1296 committed:
Не проходят тесты, которые проверяют функцию CalculatMonthsTask. Тесты для остальных функций отрабатываю правильно

Mary-cross1296 committed:
Тесты для всех функций проходят успешно, рабочая версия

Mary-cross1296 committed:
Убран отладочный код

Mary-cross1296 committed:
Исправлена функция HttpServer. Обработчики описаны в отдельных функциях

Mary-cross1296 committed:
Добавлены новые тестовые сценарии, исправлены ошибки связанные с этим

Mary-cross1296 committed:
Добавлена новая функция TentativeNextDate для предварительного расчета nextDate

Mary-cross1296 committed:
Добавлена новая функция PreliminaryNextDate для предварительного расчета nextDategi

Mary-cross1296 committed:
Добавлена функция NowBeforeNextDate, функции CalculatDayOfMonthTask и CalculatMonthsTask сокращены

Mary-cross1296 committed:
Добавлен AddTaskHandle. Тесты addtask_4_test проходят успешноr

Mary-cross1296 committed:
Изменен addtask_4_test.go

Mary-cross1296 committed:
Merge branch 'develop' of github.com:Mary-cross1296/go_final_project into develop

Mary-cross1296 committed:
Исправлена функции NowBeforeNextDat, CalculatDayOfMonthTask, CalculatMonthsTask

Mary-cross1296 committed:
Дополнительная проверка в функции NowBeforeNextDate

Mary-cross1296 committed:
Добавлен роутер, исправлены несколько названий

Mary-cross1296 committed:
Добавлен новый обработчик GetListUpcomingTaskaHandler, тесты tasks_5_test.go завершаются успешно(параметр search не учтен

Mary-cross1296 committed:
Фикс багов, все тесты работают корректно

Mary-cross1296 committed:
Фикс багов, все тесты работают корректно

Mary-cross1296 committed:
Добавлен, но не дописан обработчик редактирования задач

Mary-cross1296 committed:
Добавлены обработчики GetTaskForEdit, SaveEditTask. Тесты task_6_test и всё остальные проходят успешно

Mary-cross1296 committed:
Фикс ошибки в обработчике GetTaskForEdit. На фронте не отдавалось описание задачи

Mary-cross1296 committed:
Добавлен обработчик для удаления задач DeleteTaskHandler(). Тесты проходят успешно

Mary-cross1296 committed:
Сделана аутентификация пользователя

Mary-cross1296 committed:
Проект полностью работает локально, проходят все тесты. Из контейнера запускается, но не работают оьработчики, которые требуют авторизации пользователя и не работают тесты

Mary-cross1296 committed:
Докер файл и описанная инструкция

Mary-cross1296 committed:
Готовый докер файл и readme

Mary-cross1296 committed:
Форматирование Readme

Mary-cross1296 committed:
Форматирование Readme

dirtymew commented on Dockerfile:
```diff
@@ -0,0 +1,37 @@
+# Используем базовый образ Ubuntu
+FROM ubuntu:latest
```
я бы тут рекомендовал взять go-ный образ и сразу с той версие go что в go.mod

Mary-cross1296 replied:
+

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on Dockerfile:
```diff
@@ -0,0 +1,37 @@
+    golang-go \
+    ca-certificates \
+    curl \
+    sqlite3
```
а зачем он в контейнере?

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on Dockerfile:
```diff
@@ -0,0 +1,37 @@
+    TODO_WEB_DIR="/app/web"
+
+# Открываем порт веб-сервера
+EXPOSE 7540
```
если TODO_PORT - поменяется то экспозиться ведь продолжит 7540

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on Dockerfile:
```diff
@@ -0,0 +1,37 @@
+EXPOSE 7540
+
+# Запускаем сервер
+CMD ["/bin/sh", "-c", "/app/backend/main"]
```
а зачем через sh запускать - кажется можно и так

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on README.md:
```diff
@@ -1,5 +1,59 @@
+    `go build`
+
+    - запустить скомпилированный исполняемый файл командой:
+    `./backend.exe`
```
ну это только под windows - а программы на go тем более сервисы врядли под ним будут запускаться - тем более у тебя все docker)

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on README.md:
```diff
@@ -1,5 +1,59 @@
+
+    - ввести пароль соответсвующий паременной окружения TODO_PASSWORD из докерфайла 
+
+    - открыть консоль и во вкладке приложение открыть пункт "Файлы cookie". Скопировать значение куки token
```
тесты падают - должны проходить - это критерий приемки работы

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on README.md:
```diff
@@ -1,5 +1,59 @@
+
+    - открыть консоль и во вкладке приложение открыть пункт "Файлы cookie". Скопировать значение куки token
+
+    - вставить полученное значение в файл settings.go для переменной Token
```
с тестами я бы поступил так - все что нужно уже бы подготовил в settings.go чтобы не менять

добавил бы дефолтные значения при запуске так чтобы и бд создавалась в нужно папке(проще всего в корне проекта) и токен был готов - под дефолный пароль И так далее

чтобы просто можно было локально запустить сервер и тесты - и ничего не настраивать

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+package main
```
Тут нужен большой рефакторинг - основанная проблема что все в одном файле

почему это плохо:
 - не читаемо - соответвенно проблемы для поддержке и для изменениний
 - связность компонент и кода
 - сложно писать тесты

вот материалы которые могут помочь
 - https://github.com/golang-standards/project-layout - как "принято" органозовывать проекты
 - https://medium.com/@shershnev/layered-architecture-implementation-in-golang-6318a72c1e10 немного чистой архитектуры


думаю стоит как минимум выделить несколько пакетов 
 - storage - все что работает с бд
 - api - хендоеры веб сервера
 - date - все что с датами
 - auth - авторизация
можно еще добавить уровень сервиса для логики работы с задачами и валидаторы - но мб сходу будет не очень понятно

в задании про посылки можно подсмотреть как сделан был storage

Mary-cross1296 replied:
Вцелом поправила, наверно стало лучше ✅

dirtymew replied:
да стало лучше

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+
+// Task представляет задачу
+type Task struct {
+	//ID      int64  `json:"id"`
```
убрать лишние ненужные коменты

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+
+type Scheduler struct {
+	ID      int
+	Date    time.Duration
```
почему Duration ? а не время?

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+	_ "modernc.org/sqlite"
+)
+
+type Scheduler struct {
```
не очень понимаю зачем эта структура и почему Task недостаточно

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+	Password string `json:"password"`
+}
+
+var jwtKey = []byte("final_progect_go") // секретный ключ для подписи JWT
```
final_progect_go - это в константу стоит вынести

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+	repeat := r.FormValue("repeat")
+
+	// Преобразуем параметр "now" в формат time.Time
+	now, err := time.Parse("20060102", nowTime)
```
20060102 - в константу - много где используется

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+	w.WriteHeader(statusCode)
+	response, _ := json.Marshal(errResp)
+	w.Write(response)
```
_ , _ = w.Write(resp)

тут нужно либо явно проигнорировать ошибку или ее логировать

хорошие линтеры на это ругаться будут и не зря

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+	// Выполняем запрос INSERT в базу данных
+	tableName := "scheduler.db"
+	db, _ := OpenDataBase(tableName)
+	query := "INSERT INTO scheduler (date, title, comment, repeat) VALUES (?, ?, ?, ?)"
```
хендлеры перегружены логикой у тебя они все делают,  а им стоит заниматься свое задачей - обрабатывать запрос и отдавать ответ 

а тут и валидация и работа с бд - так они пухнут сильно их потом не оттестируешь не изменить итд

Mary-cross1296 replied:
*Разобраться

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+
+	// Выполняем запрос INSERT в базу данных
+	tableName := "scheduler.db"
+	db, _ := OpenDataBase(tableName)
```
это плохо - на каждый запрос открывать подключение к бд
- они ограничены - стоит при запуске инициировать одно подключение и его использовать

опять отправляю к заданию про посылки...

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+}
+
+// Обрабатчик Get-запросов на получение списка ближайших задач
+func GetListUpcomingTasksHandler(w http.ResponseWriter, r *http.Request) {
```
стоит с названиями поработать - слишком сложные и не понятные

просто GetTasks например ?

Mary-cross1296 replied:
Исправила названия некоторых обработчиков ✅

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+	w.Write(responseId)
+}
+
+// Обрабатчик Get-запросов на получение списка ближайших задач
```
отлично помогают комментарии ) но лучше назвать функцию хорошо и тогда возможно их не нужно будет писать

Mary-cross1296 replied:
Всё таки оставила комментарии, в название не получается поместить какие-то нюансы описания обработчика

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+			defer rows.Close()
+		} else {
+			// Если дата не распознана, делаем запрос по подстроке в title или comment
+			rows, err = db.Query("SELECT id, date, title, comment, repeat FROM scheduler WHERE title LIKE ? OR comment LIKE ? ORDER BY date DESC LIMIT 50", "%"+search+"%", "%"+search+"%")
```
лиммит стоит в константу вынести

стоит использовать именновыные параметры чтобы не дублировать "%"+search+"%", "%"+search+"%"

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+			}
+			defer rows.Close()
+		}
+	} else {
```
тут стоит поработать с ветлением - else зло лучще его не использовать по возможности

тут switch напрашивается

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+		task.ID = fmt.Sprint(id)
+		tasks = append(tasks, task)
+	}
+
```
нужно еще ошибку из rows обработать
http://go-database-sql.org/errors.html

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+		}
+	}
+
+	// Проверка существования задачи перед обновлением
```
это лишнее и не оптимально - 2 запросы там где можно одним обойтись

если нужно проверить что UPDATE сработал и обновил что-то  - то просто нужно проверь RowsAffected

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+
+	var task Task
+	var id int64
+	err := db.QueryRow("SELECT id, date, title, comment, repeat FROM scheduler WHERE id = ?", idParamNum).Scan(&id, &task.Date, &task.Title, &task.Comment, &task.Repeat)
```
тут вот как раз логика для сервиса - если вынести CRUD операции в отдельные методы то такую логику можно было просто их комбинацией организовать Get/Update/Delete

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+	}
+
+	// Проверка количества затронутых строк
+	rowsAffected, err := result.RowsAffected()
```
ну вот - почему с update так не сделать?

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+	}
+
+	// Чтение тела запроса
+	fmt.Printf("Отладка TODO_PASSWORD: %s\n", os.Getenv("TODO_PASSWORD"))
```
тут тоже не хорошо
 - во первых отладочный принт - его стоит убрать
 - во вторых есть же log у тебя - стоит что-то одно использовать
 - в третьих это syscall к ОС за переенной окружения - все переменные окружения стоит во время запуска приложения прочитать и больше не ходить за ними

Mary-cross1296 replied:
1 - Исправила ✅
2 - Исправила ✅
3 - Исправила ✅

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+
+func auth(next http.HandlerFunc) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		pass := os.Getenv("TODO_PASSWORD")
```
уже писал про это выше

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+	}
+
+	// Запуск сервера на указанном порту
+	log.Printf("Сервер запущен на порту %v\n", port)
```
супер что добавила логирование

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+func ChekingDataBase() error {
+	tableName := "scheduler.db"
+
+	appPath, err := os.Executable()
```
тут лучше использовать относительные пути - Getwd например

os.Executable() на разных ОС по разному работают

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+	return counterNegativeNum
+}
+
+func FindMinNum(nums []int, numNegative int) int {
```
не знаю зачем но если что в go в новых версиях завезли функици max и min

Mary-cross1296 replied:
Оставила как есть

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,1160 @@
+	// Создание канала для поступления сигналов
+	sigs := make(chan os.Signal, 1)
+	var sig os.Signal
+	for {
```
ой - это не очень хорошо - простой способ повиснуть

если хочешь сделать gracefull shutdown (я считаю что тут это не нужно делать - по тому как сделать его правильно это прям сильно за рамками базового курса)

то просто запустить в горутине с каналом или waitgroup ... 

ни или жди чего-то из sigs

Mary-cross1296 replied:
Оставила как есть

dirtymew replied:
@Mary-cross1296 тут прям плохо сделано - я бы убрал это все

gracefull shutdown корректно сделать не так просто - это сильно за рамками курса

уже писал про лишний бесконечный цикл

на syscall надо подписываться до запуска горутины с сервером и еще много чего тут не так

если прямо очень хочется то в интеренете куча статей как про это - там можно подсмотреть...   но сейчас реализация некорректная

Mary-cross1296 replied:
Исправила ✅

dirtymew reviewed (changes requested):
Привет Мария

хорошая работа - отлично что добавила логирование и сделала задания со звездочкой

есть конечно что улучшить и исправить  - основные моменты:
 - запуск/настройка и прохождение тестов
 - все в одном файле - стоит разбить на несколько пакетов/файлов

Mary-cross1296 committed:
Исправлена логика работы для обработчика AddTaskHandler().Все тесты проходят успешно

Mary-cross1296 committed:
Тесты проходят из докер контейнера

Mary-cross1296 committed:
Добавлено автоматическое получение Токена для наиболее простого запуска тестов

Mary-cross1296 committed:
Исправлены названия некоторых переменных

Mary-cross1296 committed:
Рефакторинг проекта.

Mary-cross1296 committed:
Изменение докер файла

Mary-cross1296 committed:
Добавлен файл с переменными окружения. Обновлен Dockerfile, теперь тесты проходят успешно локально и из конейнера.

Mary-cross1296 committed:
Добавить чтение переменных окружение и их запись в глобальные переменные.
Добавить использование относительного пути при проверке БД.
Удалить отладочный и закомментированный код.

Mary-cross1296 committed:
Исправлены все замечания + отладочный код. Тесты локально и в докере проходят.

Mary-cross1296 committed:
Исправлены замечания. Убран отладочный код. Тесты в докере и локально проходят успешно

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,84 @@
+	server := api.HttpServer(port, webDir, Db)
+
+	// Получение и обновление токена
+	if err := utils.GetAndUpdateToken(); err != nil {
```
это тоже плохо 

тесты и код вещи не зависимые - не нужно ничего в тестах обновлять

все просто - сделай один токен для тестов с деволтный паролем и все

тут сильно перемудрила...

Mary-cross1296 replied:
Не очень поняла, как правильно реализовать это.
Поэтому прописала дефолтное значение токену в settings.go, затем в функции Auth добавила проверку на дефолтный токен

Исправила ✅

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,84 @@
+)
+
+// Глобальная переменная для подключения к базе данных
+var Db *sql.DB
```
зачем глобальная переменная?  она совершенно не нужна

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on Dockerfile:
```diff
@@ -0,0 +1,23 @@
+RUN go mod download
+
+# Компилируем программу
+RUN CGO_ENABLED=1 GOOS=linux go build -o /app/backend/main ./backend/main.go
```
лучше CGO_ENABLED=1 GOOS=linux в ENV

Mary-cross1296 replied:
Исправила ✅

dirtymew replied:
нет. ты добавила их в ENV а в RUN оставила

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on dateCalc/calculations.go:
```diff
@@ -0,0 +1,241 @@
+package dateCalc
```
не очень название для файла и пакета

мб просто dates/task.go и dates.go

Mary-cross1296 replied:
Исправила ✅

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on dateCalc/utils.go:
```diff
@@ -0,0 +1,129 @@
+package dateCalc
```
utils/tools/helpers -  считаются плохой практикой

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on api/server.go:
```diff
@@ -0,0 +1,53 @@
+	go func() {
+		err := httpServer.ListenAndServe()
+		if err != nil && err != http.ErrServerClosed {
+			log.Fatal("Http server error \n", err)
```
вот тут вылетает паника или просто завершается по ошибке

далее у тебя заблоченный на канал бескочнный цикл в рутине - пока syscall не придет сервис будет висеть .... 

подумай - если не понимаешь как правильно сделать - лучше убери

Mary-cross1296 replied:
Убрала обработчик сигналов из main.go 
Исправила ✅

Mary-cross1296 replied:
Вот это Исправила ✅, но не до конца понимаю смысл
Просто исправила log.Fatal на log.Printf

dirtymew commented on auth/auth.go:
```diff
@@ -0,0 +1,62 @@
+	"golang.org/x/crypto/bcrypt"
+)
+
+const JwtKey = "final_progect_go" // секретный ключ для подписи JWT
```
мб тоже в ENV ?

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on tests/addtask_4_test.go:
```diff
@@ -148,6 +149,8 @@ func TestAddTask(t *testing.T) {
 				continue
 			}
 			if today && task.Date != now.Format(`20060102`) {
+				fmt.Printf("today %v \n", today)
```
тут зачем-то тесты потрогала - откати

Mary-cross1296 replied:
Исправила ✅, откатила тесты в этом файле

dirtymew commented on tests/app_1_test.go:
```diff
@@ -21,17 +21,26 @@ func getURL(path string) string {
 func getBody(path string) ([]byte, error) {
-	resp, err := http.Get(getURL(path))
+	url := getURL(path)
+	//fmt.Printf("Запрашиваемый URL: %s \n", url) // Отладочный вывод
```
убрать лишние комменты

Mary-cross1296 replied:
Исправила ✅

dirtymew reviewed (changes requested):
Привет Мария

есть что еще подправить и улучшить 

дал комментарии

Mary-cross1296 committed:
Рабочая версия после исправления замечаний. Добавить дефолтный пароль и токен для тестов

Mary-cross1296 committed:
Исправлены все замечания после ревью

Mary-cross1296 committed:
Тесты локально и в докере проходят успешно

dirtymew commented on api/handlers.go:
```diff
@@ -0,0 +1,536 @@
+	switch {
+	case search == "":
+		// Делаем запрос по поиску всех задач с сортировкой по дате
+		rows, err = h.DB.Query("SELECT id, date, title, comment, repeat FROM scheduler ORDER BY date DESC LIMIT ?", MaxTasksLimit)
```
это все конечно лучше бы в storage 

 задании про посылки - можно посмотреть как сделан сервис и сторадж

тут тоже самое стоило повторить только еще сверху слой handler-ов над сервисом

так очень перегруженые хенжлеры получились - код не абсрактный - плохо расширяемый, тестировать это будет невозможно..

dirtymew replied:
если не хочешь - можешь не рефакторить, но вместо заданий со звездочкой я бы время на это потратил лучше

Mary-cross1296 replied:
Исправила ✅

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,54 @@
+	}
+
+	// Ожидание завершения работы сервера
+	select {}
```
чесно сказать выглядит не очень - не понятно зачем... почему просто  httpServer.ListenAndServe() не вернуть?

Mary-cross1296 replied:
Исправила ✅

dirtymew reviewed (changes requested):
Привет Мария

дал комментарии - все не критичное, если не будешь рефакторить, то верни обратно я готов принять работу.

Mary-cross1296 committed:
Запросы к базе данных выделены в отдельные функции. Тесты локально и из контейнера проходят успешно

Mary-cross1296 committed:
Доработан запуск и остановка сервера

dirtymew commented on storage/dbQuery.go:
```diff
@@ -0,0 +1,151 @@
+
+	if rowsAffected == 0 {
+		log.Printf("DeleteTaskHandler(): Task not found")
+		return err
```
тут помарка - возвращать не буду @Mary-cross1296 обрати внимание

err тут nil

поэтому нужно вернуть свою ошибку errors.New("not found")

dirtymew commented on backend/main.go:
```diff
@@ -0,0 +1,58 @@
+package main
```
backend/backend.exe не стоит коммитеть

dirtymew reviewed (approved):
Привет Мария

получилось неплохо 

Поздравляю и желаю удачи
